pkg/edit/navigation_test.go

```diff
@@ -63,23 +66,91 @@ func TestNavigation_WidthRatio(t *testing.T) {
 	f.TestTTY(t,
-		"~"+string(os.PathSeparator)+"d> ", term.DotHere, "\n",
+		filepath.Join("~", "d"), "> ", term.DotHere, "\n",
 		" NAVIGATING  \n", Styles,
 		"************ ",
-		" d                a               ", Styles,
+		" d                a               \n", Styles,
 		"################ ++++++++++++++++ ",
+		"                  e              ", Styles,
+		"                 ////////////////",
+	)
+}
+
+// Test some corner cases:
+//
+// 1) Inserting a selection when the CLI cursor is at the start of the edit
+// buffer omits the space char prefix.
+//
+// 2) Inserting the "selection" in an empty directory inserts nothing rather
+// than the literal ` ''`  that used to be inserted.
+//
+// 3) Inserting a selection when the CLI cursor is not at the start of the
+// edit buffer but the preceding char is a space does not prepend another
+// space.
+//
+// See https://github.com/elves/elvish/issues/1169
+func TestNavigationCornerCases(t *testing.T) {
+	f, cleanup := setupNav()
+	defer cleanup()
+
+	// Validate that inserting a file name at the start of the command line
+	// does not prepend a space.
+	f.TTYCtrl.Inject(term.K('N', ui.Ctrl))
+	f.TTYCtrl.Inject(term.K(ui.Enter, ui.Alt))
+	f.TestTTY(t,
+		filepath.Join("~", "d"), "> ",
+		"a", Styles,
+		"!", term.DotHere, "\n",
+		" NAVIGATING  \n", Styles,
+		"************ ",
+		" d      a                 \n", Styles,
+		"###### ++++++++++++++++++ ",
+		"        e                ", Styles,
+		"       //////////////////",
+	)
+
+	// Validate that inserting an empty directory is a nop.
+	f.TTYCtrl.Inject(term.K(ui.Down))          // select "e"
+	f.TTYCtrl.Inject(term.K(ui.Right))         // move into the empty "e" directory
+	f.TTYCtrl.Inject(term.K(ui.Enter, ui.Alt)) // insert nothing since the dir is empty
+	f.TestTTY(t,
+		filepath.Join("~", "d"), "> ",
+		"a", Styles,
+		"!", term.DotHere, "\n",
+		" NAVIGATING  \n", Styles,
+		"************ ",
+		" a                        \n", Styles,
+		"                          ",
+		" e    ", Styles,
+		"######",
+	)
+
+	// Validate that inserting a file name when the previous char in the buffer is
+	// a space does not insert another space.
+	f.TTYCtrl.Inject(term.K(' '))              // user preses [space]
+	f.TTYCtrl.Inject(term.K(ui.Left))          // select "e"
+	f.TTYCtrl.Inject(term.K(ui.Enter, ui.Alt)) // insert the "e" file name
+	f.TestTTY(t,
+		filepath.Join("~", "d"), "> ",
+		"a e", Styles,
+		"!", term.DotHere, "\n",
+		" NAVIGATING  \n", Styles,
+		"************ ",
+		" d      a                 \n", Styles,
+		"######                    ",
+		"        e                ", Styles,
+		"       ##################",
 	)
 }
 
 func setupNav() (*fixture, func()) {
 	f := setup()
 	restoreLsColors := lscolors.WithTestLsColors()
 
-	testutil.ApplyDir(testutil.Dir{"d": testutil.Dir{"a": ""}})
-	err := os.Chdir("d")
-	if err != nil {
-		panic(err)
-	}
+	// Path "d/a" is an empty file and "d/e" is an empty directory.
```

Comment:
There is no need to change this.

Reply:
I beg to differ and feel pretty strongly about this aspect of my change. The `testutil.ApplyDir` function is clever but not clear. When I made this change I couldn't see how to use it to create an empty directory. I now see that should be possible by doing `"e":  testutil.Dir{}`. Ugh! I actually want to see all uses of `testutil.ApplyDir()` replaced by their obvious counterparts. I can't see any justification for this API given it makes it harder for people to understand what a unit test is doing. Especially given that they already need to understand what functions like `testutil.MustMkdirAll` and `testutil.MustCreateEmpty` do. In that sense `testutil.ApplyDir` is nothing more than additional cognitive load with no real benefit.

Reply:
Coincidentally I watched https://www.youtube.com/watch?v=NwEuRO_w8HE, "Opening keynote: Clear is better than clever - GopherCon SG 2019" just after writing my previous comment since it popped up in my Youtube recommendations. It seems applicable to this discussion.

Reply:
I don't see a strong objective argument against `testutil.ApplyDir`. Please keep it. If you'd like to discuss this further please raise the issue separately.

Reply:
@xiaq, I see that you introduced `testutil.Dir` and `testutil.ApplyDir` a year ago. While that API has relatively few uses at this time it seems clear you intend it to be the default, most often used, API for creating unit test directory trees. Most of my objection to this relatively new API, in the context of this change, is because the existing uses are a) rare, and b) formatted in a fashion that actually obscured their behavior. I just pushed an amendment to this P.R. that should resolve this aspect of my P.R. and illustrate what I think is a clearer expression of the `testutil.Dir` pattern.